### PR TITLE
Fix focus loop in portal

### LINKED ISSUE #2168: Focus loop crashes the browser
### Current behavior

Using the keyboard navigation to open a Popover with no focusable elements causes the browser to crash.

### Steps to reproduce the bug

1. Open sandbox: https://codesandbox.io/s/clever-river-9rfl7w?file=/src/popover.tsx
2. Press the tab to focus on the button
3. Press enter to open the popover
4. Press the tab again
5. The browser freezes

### Expected behavior

The browser should not crash when no other element is available.

### Workaround

https://github.com/ariakit/ariakit/issues/2168#issuecomment-1366314645

### Possible solutions

This is probably related to the `preserveTabOrder` logic:

https://github.com/ariakit/ariakit/blob/7a575278ef438fdeb7f43859a67f14f6e31f7f9a/packages/ariakit/src/portal/portal.tsx#L177-L244

When there's no tabbable element, the focus traps likely move focus between themselves continuously, causing an infinite loop.

### PR DESCRIPTION
**Pull request by georgekaran**
Closes #2168

### CODE REVIEW

**Comment by diegohaz** [comment]
Thank you for working on this @georgekaran! Can we have an example with a test to guarantee this won't break in the future?

**Comment by georgekaran** [comment]
> Thank you for working on this @georgekaran! Can we have an example with a test to guarantee this won't break in the future? It can be a popover example.

Sure thing, let me know if any changes are necessary.

File: packages/ariakit/src/popover/__examples__/popover-portal/index.tsx [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { Button } from "ariakit/button";
+import { Popover, usePopoverState } from "ariakit/popover";
+import "../popover/style.css";
```
Comment (diegohaz) [inline packages/ariakit/src/popover/__examples__/popover-portal/index.tsx]: [This doesn't work](https://ariakit-git-fork-georgekaran-fix-portal-focus-loop-ariakit.vercel.app/examples/popover-portal). You have to create a `style.css` file inside this folder and `@import url("../popover/style.css")` from there.
```suggestion
import "./style.css";
```

File: packages/ariakit/src/popover/__examples__/popover-portal/test.tsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { getByRole, press, render } from "ariakit-test";
+import Example from ".";
+
+const getPopover = () => getByRole("dialog", { hidden: true });
+
+const getDisclosure = () => getByRole("button", { name: "Open" });
+test("navigation when using keyboard on a popover with portal", async () => {
```
Comment (diegohaz) [inline packages/ariakit/src/popover/__examples__/popover-portal/test.tsx]: Nit:
```suggestion
const getPopover = () => getByRole("dialog", { hidden: true });
const getDisclosure = () => getByRole("button", { name: "Open" });

test("navigation when using keyboard on a popover with portal", async () => {
```

File: packages/ariakit/src/portal/portal.tsx [hunk 3]
```diff
@@ -181,7 +181,13 @@ export const usePortal = createHook<PortalOptions>(
                 ref={beforeInsideRef}
                 onFocus={(event) => {
                   if (isFocusEventOutside(event, portalNode)) {
-                    queueFocus(getNextTabbable());
+                    const nextFocusableElement = getNextTabbable();
+
+                    // When no other focusable element is found, prevent the focus trap
+                    // to focus on each other which leads to an infinite loop.
+                    if (nextFocusableElement !== afterInsideRef.current) {
```
Comment (diegohaz) [inline packages/ariakit/src/portal/portal.tsx]:
```suggestion
                    // When no other focusable element is found, prevent the focus trap
                    // to focus on each other which leads to an infinite loop.
                    // See https://github.com/ariakit/ariakit/issues/2168
                    if (nextFocusableElement !== afterInsideRef.current) {
```

**Review by diegohaz — APPROVED** [review]
Nice! Thank you @georgekaran!

**Comment by agriffis** [comment]
Is it possible that this didn't fix the entire issue? I'm running 2.0.0-next.43 and hitting infinite loops as soon as I render the dialog in my application, without even needing to click it open.

**Pull request merged**